DESIGN.md

```diff
@@ -0,0 +1,111 @@
+
+
+## Architectural Design
+The design of this project will center around the modifications to SqlTable. The design for other components in the storage layer will remain unchanged. On a schema change the SqlTable will create a new DataTable that will store all the tuples inserted from that point on into the new version. To be lazy it will not modify already existing tuples to transform them into the latest version.  In order to support this lazy schema change we need two functionalities: maintaining tuples in multiple different schema versions and providing methods of transforming them into the desired version.
+
+### Multi-versioning
+To address the multi-versioning the SqlTable will maintain a map from the schema_version_number to a DataTable. There will be one DataTable for each schema version. The functionality for accessing tuples is already present in DataTable and this way SqlTable will only need to manage the two functionalities we described above. The rest of the functionalities will be handled by the already existing DataTable implementation. Furthermore, each block will maintain metadata of which version all of the tuples within the block belong to. Since each DataTable can only be a single version, a block cannot contains tuples from multiple versions.  Below is a description of the multiversion design for each of the SqlTable operations we are modifying and we will refer to expected version as the version of the tuple the user expects to see and actual version as the version the tuple is currently in.
+
+#### SqlTable::Insert
+Insert will always insert the passed in tuple into the DataTable of the schema version number passed in
+
+#### SqlTable::Update
+- Update has three cases
+    1. The expected schema version matches the actual schema version
+        - The update will happen in place on the DataTable of the actual schema version
+    2. The expected schema version doesn’t match the actual schema version but the update doesn’t touch any columns not in the actual schema version
+        - The update will happen in place on the DataTable of the expected schema version
+    3. The expected schema version doesn’t match the actual schema version and the update touches not in the actual schema version. The following steps occur
+        - Retrieve the tuple from the actual version DataTable
+        - Transform the tuple to the expected version
+        - Delete the tuple from the actual version DataTable
+        - Insert the tuple into the expected version DataTable
+
+#### SqlTable::Scan
+- SqlTable will maintain its own slot iterator which will be used to iterate across all the schema version. The iterator interface exposed to the user will not change
+- The iterator for Scan must always begin on the latest version
```

Comment:
I am doubting if this method really solves every problem. See the trace in https://github.com/cmu-db/terrier/pull/342/files#r275173692

Reply:
As the discussion in https://github.com/cmu-db/terrier/pull/342/files#r275175240, it seems it is not the actual solution you used to solve the scan/update conflict. You rely on MVCC to solve the conflict. Maybe the document also need to change accordingly.

Reply:
This requirement is solving a different problem than what was posed there.  MVCC solves the problem of a race across transactions during the insert/delete process, but this addresses the risk of processing the same logical tuple twice.  Specifically, if we are using a Hyper-style execution model of process a group of tuples as far as we can and this pipeline involves an update that forces migration to the latest version, then we **must** process the latest version first.

For example, if we add a column and then systematically set the value of this column (via update and not a default value at the time we changed the schema), then we will migrate every tuple to the current version.  If we iterate over older versions first, then when we reach the current version they will be visible because a transaction can see its own writes.  While we could handle this by tracking migrated tuples explicitly, it is cleaner to iterate over the current version first so that when we trigger migrations they are inserted behind the `SlotIterator` and won't be visited again.

Reply:
I see. So the user must guarantee that each update/scan call does not interleave with each other. But they can call scan, then update, then scan, etc, which is solved by this method.

Reply:
Does it mean we are only allowed to update the scanned tuples in the same transaction? For example, if we updated an unscanned tuple and it was migrated to an already scanned data table, then we would never be able to see it in the scan.

Reply:
If the update occurred in a different transaction, then the scan will see the old "deleted" tuple because the other transactions writes are not visible to the scan (snapshot isolation).